src/hotspot/cpu/riscv/vm_version_riscv.cpp

```diff
@@ -181,6 +183,51 @@ void VM_Version::initialize() {
     }
   }
 
+  if (UseZba && !_cpu_features.ext_zba) {
```

Comment:
Unlike RVC, there is still no hardware that supports those extensions. I think we should do more testing and evaluation on real hardware before enabling them by default.

Reply:
QEMU has good enough support to do functional testing. For performance testing, I agree it would be ideal to have proper hardware to test, but even then it's not guaranteed that a performance-minded extension would _actually_ be faster on all given hardwares. For example the original AVX-512 which led to thermal throttling.

So IMO as long as it's functional, we should want to enable it. We otherwise run the risk of "what about"-ing it until we can never enable it.

Happy to heard others' opinions as well!

Reply:
That's a good question. I am thinking that at least the user could enable use of specific ISA extensions on the JVM command line if it turns out to be profitable on their hardware platforms. It should be safe for them to do as this PR adds the necessary detection of availability of those ISA extensions. Previous working experience on other CPU platforms show that it's might be risky in respect of performance for the JVM to auto-enable all those [1][2]: the performance numbers might vary on different vendors. So I guess one possibility is that we might want some vendor-specific control like we do other CPU platforms [3]. But I hope some ISA extensions like bitmanip could be generally profitable across all venders so that the JVM could auto-enable those extension. But that needs more testing on real hardwares to prove. That said, we could at least move JVM options corresponding to those extensions out of the EXPERIMENTAL group for now once they are fully tested in QEMU system, but that's another new issue.

[1] https://bugs.openjdk.org/browse/JDK-8292894
[2] https://bugs.openjdk.org/browse/JDK-8295698
[3] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/vm_version_aarch64.cpp#L132

Reply:
> It should be safe for them to do as this PR adds the necessary detection of availability of those ISA extensions.

Ok. That also aligns us with how it's done on other platforms (SHA on x86 or AArch64 for example).

I've a slight concern with the reliability of detecting these features based solely on the device tree + being able to force enable said-features in case we know for sure the extension is available but it's not communicated by the kernel.

FYI, the kernel doesn't transfer verbatim the device-tree "isa" provided by the hardware, but it reconstruct it based on the extensions the kernel knows about. So if the hardware support a cutting edge extension (Zvknh and Zvkb for example), but the kernel is older and doesn't know about it, the hardware is going to provide a device-tree with `_Zvknh_Zvkb` to the kernel, but it's not going to show up in `/proc/cpuinfo`. (You can find the list of extensions the kernel currently knows about at [1], and the logic to build the `isa` string at [2]).

> So I guess one possibility is that we might want some vendor-specific control like we do other CPU platforms

I fully agree we will want some of these. Good thing there are the vendorid as part of the spec so we can easily detect which one is which.

> But I hope some ISA extensions like bitmanip could be generally profitable across all venders so that the JVM could auto-enable those extension. But that needs more testing on real hardwares to prove.

Agreed. As long as we have a way to manually enable them so that the user still has the ability to test out some features on their hardware.

[1] https://github.com/torvalds/linux/blob/0983f6bf2bfc0789b51ddf7315f644ff4da50acb/arch/riscv/kernel/cpu.c#L164-L172
[2] https://github.com/torvalds/linux/blob/0983f6bf2bfc0789b51ddf7315f644ff4da50acb/arch/riscv/kernel/cpu.c#L174-L191

Reply:
> > It should be safe for them to do as this PR adds the necessary detection of availability of those ISA extensions.
> 
> Ok. That also aligns RISC-V with how it's done on other platforms (SHA on x86 or AArch64 for example).
> 
> I've a slight concern with the reliability of detecting these features based solely on the device tree + being able to force enable said-features in case we know for sure the extension is available but it's not communicated by the kernel.
> 
> FYI, the kernel doesn't transfer verbatim the device-tree "isa" provided by the hardware, but it reconstruct it based on the extensions the kernel knows about. So if the hardware support a cutting edge extension (Zvknh and Zvkb for example), but the kernel is older and doesn't know about it, the hardware is going to provide a device-tree with `_Zvknh_Zvkb` to the kernel, but it's not going to show up in `/proc/cpuinfo`. (You can find the list of extensions the kernel currently knows about at [1], and the logic to build the `isa` string at [2]).

That make sense to me. My initial thought was to have some sort of checking like this PR does when users try to enable use of isa extensions on the command line. It will be better than simply crashing the JVM when their hardware are not equipped with those features. But as you mentioned, this might also bear the risk of losing the ability for them to do so. In the hope of having some stable user interface for probing riscv hardware features like [1], I guess we might stay with the current approach and put off this PR for some while to see.

[1] https://lkml.org/lkml/2023/2/6/1147

Reply:
> I guess we might need to stay with the current approach and put off this PR for some while to see
Agree.  Looks like it's still too early to do that. I'll convert this PR to draft then.

Reply:
An alternative approach that doesn't rely on a kernel API is to generate some stub that probes for some instructions, and trigger a SIGILL in case the instruction isn't supported. That SIGILL can be caught by the runtime (very similar to how we do it with SIGSEGV and SafeFetch for example) and we know based on the instruction executed that it isn't available.

That would work as long as there is no conflicts in the encoding of instructions, and no bug/performance penalty in the actual implementation of the instructions (but that's where vendor-specific settings is helpful).